packages/next-on-pages/src/buildApplication/processVercelFunctions/ast.ts

```diff
@@ -0,0 +1,267 @@
+import type { Node } from 'acorn';
+import type * as AST from 'ast-types/gen/kinds';
+import type { ProcessVercelFunctionsOpts } from '.';
+import { cliError } from '../../cli';
+
+/**
+ * Collects the identifiers from the AST and adds them to the provided maps and lists.
+ *
+ * @param collectedIdentifiers The collected identifiers.
+ * @param program The AST for the function file.
+ * @param opts The options for the function processing.
+ */
+export function collectIdentifiers(
+	{ identifierMaps, programIdentifiers }: CollectIdentifiersOpts,
+	program: AST.ProgramKind,
+	{ disableChunksDedup }: Partial<ProcessVercelFunctionsOpts>,
+): void {
+	// Wasm
+	collectIdentifierType(
+		{ programIdentifiers, foundIdentifiers: identifierMaps.wasm },
+		program,
+		extractWasm,
+	);
+
+	// Manifests
+	collectIdentifierType(
+		{ programIdentifiers, foundIdentifiers: identifierMaps.manifest },
+		program,
+		extractManifest,
+	);
+
+	// Webpack chunks
+	if (!disableChunksDedup) {
+		collectIdentifierType(
+			{ programIdentifiers, foundIdentifiers: identifierMaps.webpack },
+			program,
+			extractWebpack,
+		);
+	}
+}
+
+type CollectIdentifiersOpts = {
+	identifierMaps: Record<IdentifierType, IdentifiersMap>;
+	programIdentifiers: ProgramIdentifiers;
+};
+
+/**
+ * Collects a specific type of identifier from the AST and adds them to the provided map and list.
+ *
+ * @param collectedIdentifiers The collected identifiers.
+ * @param program The AST for the function file.
+ * @param callback A function that extracts the identifiers from a statement.
+ */
+function collectIdentifierType<T extends IdentifierType>(
+	{ foundIdentifiers, programIdentifiers }: CollectIdentifierTypeOpts,
+	program: AST.ProgramKind,
+	callback: (
+		statement: AST.StatementKind,
+	) => RawIdentifier<T> | null | RawIdentifier<T>[],
+): void {
+	const rawIdentifiers = program.body
+		.map(callback)
+		.flat()
+		.filter(Boolean) as RawIdentifier<T>[];
+
+	// Set to track for identifier collisions in a single file
+	const uniqueIdentifiers = new Set<string>();
+
+	for (const ident of rawIdentifiers) {
+		programIdentifiers.push(ident);
+
+		const existing = foundIdentifiers.get(ident.identifier);
+		if (!existing) {
+			// Identifier doesn't exist yet, add it.
+			foundIdentifiers.set(ident.identifier, { consumers: 1 });
+		} else if (!uniqueIdentifiers.has(ident.identifier)) {
+			// Identifier already exists, but it's not a collision in this file.
+			existing.consumers += 1;
+		} else {
+			// Identifier already exists and it was already found in this file - this is a collision.
+			cliError(
+				`
+						ERROR: Detected a collision with the webpack chunks deduplication.
+									 Try adding the '--disable-chunks-dedup' argument to temporarily solve the issue.
+					`,
```

Comment:
what about an error message such as this?
```suggestion
				`ERROR: A duplicated identifier has been detected in the same function file, aborting.`,
```

having the suggestion to try adding `--disable-chunks-dedup` could be beneficial, but I wouldn't want the error message to get too long, maybe this would be enough? (and if people encounter this issue and report it we can revisit this later?)